File: shortest_bridge.py
class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        queue = deque()
        visited = set()
        # Find any island and mark its cells as 2 and set starting positions
        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] != 1:
                return
            grid[r][c] = 2
            queue.append((r, c))
            visited.add((r, c))
            dfs(r+1, c)
            dfs(r-1, c)
            dfs(r, c+1)
            dfs(r, c-1)                                    

        found = False
        for r in range(rows):
            if found:
                break
            for c in range(cols):
                if grid[r][c] == 1:
                    dfs(r, c)
                    found = True
                    break

        # Perform a BFS starting from the cells marked as 2 
        distance = 0
        while queue:
            for _ in range(len(queue)):
                row, col = queue.popleft()

                for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                    new_row, new_col = row + dy, col + dx

                    if 0 <= new_row < rows and 0 <= new_col < cols and(new_row, new_col) not in visited:
                        if grid[new_row][new_col] == 1:
                            return distance
                        
                        queue.append((new_row, new_col))
                        visited.add((new_row, new_col))
            distance += 1
            
            

File: shortest_bridge.java
class Solution {
    public int shortestBridge(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        Deque<int[]> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        // Find any island and mark its cells as 2 and set starting positions
        class DFS {
            void dfs(int r, int c) {
                if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] != 1) {
                    return;
                }
                grid[r][c] = 2;
                queue.offer(new int[]{r, c});
                visited.add(r + "," + c);
                dfs(r + 1, c);
                dfs(r - 1, c);
                dfs(r, c + 1);
                dfs(r, c - 1);
            }
        }

        boolean found = false;
        DFS dfs = new DFS();
        for (int r = 0; r < rows; r++) {
            if (found) {
                break;
            }
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 1) {
                    dfs.dfs(r, c);
                    found = true;
                    break;
                }
            }
        }

        // Perform a BFS starting from the cells marked as 2
        int distance = 0;
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] cell = queue.poll();
                int row = cell[0];
                int col = cell[1];

                for (int[] dir : directions) {
                    int newRow = row + dir[0];
                    int newCol = col + dir[1];

                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols &&
                            !visited.contains(newRow + "," + newCol)) {
                        if (grid[newRow][newCol] == 1) {
                            return distance;
                        }

                        queue.offer(new int[]{newRow, newCol});
                        visited.add(newRow + "," + newCol);
                    }
                }
            }
            distance++;
        }

        return -1;        
    }
}
